File: AtCoder_Beginners_Selection/ABC087B/main.nim
import strutils

proc main =
    let A: int = parseInt(readLine(stdin)) # 500円の個数
    let B: int = parseInt(readLine(stdin)) # 100円の個数
    let C: int = parseInt(readLine(stdin)) # 50円の個数
    let X: int = parseInt(readLine(stdin)) # 合計金額

    var result: int = 0

    # input value check
    assert(0 <= A and A <= 50)
    assert(0 <= B and B <= 50)
    assert(0 <= C and C <= 50)
    assert((A+B+C) >= 1)
    assert(50 <= X and X <= 20000)
    assert(X mod 50 == 0)

    # 総当たりで確認
    for quantity_A in 0 .. A:
        for quantity_B in 0 .. B:
            for quantity_C in 0 .. C:
                if X == (500 * quantity_A + 100 * quantity_B + 50 * quantity_C):
                    inc(result)
    echo result

when isMainModule:
    main()

File: AtCoder_Beginners_Selection/ABC087B/main.py
import operator

def main():
    A: int = int(input()) # 500円の個数
    B: int = int(input()) # 100円の個数
    C: int = int(input()) # 50円の個数
    X: int = int(input()) # 合計金額
    result: int = 0

    # input value check
    assert(0 <= A <= 50)
    assert(0 <= B <= 50)
    assert(0 <= C <= 50)
    assert((A + B + C) >= 1)
    assert(50 <= X <= 20000)
    assert(operator.mod(X, 50) == 0)

    for coin_A in range(0, A+1):
        for coin_B in range(0, B+1):
            for coin_C in range(0, C+1):
                if X == 500 * coin_A + 100 * coin_B + 50 * coin_C:
                    result += 1
    print(result)

if __name__ == "__main__":
    main()
